In [3]:
import numpy as np

### 1. 넘파이를 사용하여 다음과 같은 행렬을 만들고, 이 배열에서 값 70을 인덱싱하세요.  (3.1 - 연습문제 1 변형)
10 20 30 40 <br>
50 60 70 80 <br>

In [19]:
# 밑에 코드 작성
a = np.array([[10, 20, 30, 40], [50, 60, 70, 80]])
a

array([[10, 20, 30, 40],
       [50, 60, 70, 80]])

In [27]:
a[1,2]

70

### 2. 다음 배열은 첫번째 행(row)에 학번, 두번째 행에 영어 성적, 세번째 행에 수학 성적을 적은 배열이다. 영어 성적을 기준으로 각 열(column)을 재정렬하라. (연습문제 3.3.2) # 정렬

In [31]:
v = np.array([[  1,    2,    3,    4],
       [ 46,   99,  100,   71],
       [ 81,   59,   90,  100]])

# 밑에 코드 작성

v1= np.sort(v, axis=1)

print(v1)


[[  1   2   3   4]
 [ 46  71  99 100]
 [ 59  81  90 100]]


### 3. 코사인 유사도 함수 구현하기 ★

In [60]:
def cos_sim(A,B):
    # 밑에 코드 작성
    passpass

In [60]:
def cos_sim(A,B):
    # 밑에 코드 작성
    dot_product = np.dot(A, B)
    norm_A = np.linalg.norm(A)
    norm_B = np.linalg.norm(B)
    similarity = dot_product / (norm_A * norm_B)
    return similarity

In [61]:
# Test
doc1 = np.array([0,1,1,1])
doc2 = np.array([1,0,1,1])
result = round(cos_sim(doc1, doc2),2)
result

0.67

- 정답 : 0.67

## 4,5번 문제 : 붓꽃 데이터 활용 ★

### 4. 붓꽃들의 변수별 가중치를 통해 전체 붓꽃 가중평균 구하기

- 변수별 가중치
    - 꽃받침의 길이 : 0.3, 꽃받침의 폭 : 0.2, 꽃잎의 길이 : 0.25, 꽃잎의 폭 : 0.25

In [4]:
from sklearn.datasets import load_iris

iris = load_iris() # 붓꽃 데이터 로드
X = iris.data
weights = np.array([0.3, 0.2, 0.25, 0.25])

def weighted_average(data, weights):
    
    # 밑에 코드 작성
    weighted_average = np.dot(data, weights)   # dot 내적 계산 함수
    
    return np.mean(weighted_average)   # 평균
    

result = weighted_average(X, weights)
result

3.6038

In [5]:
X

array([[5.1, 3.5, 1.4, 0.2],
       [4.9, 3. , 1.4, 0.2],
       [4.7, 3.2, 1.3, 0.2],
       [4.6, 3.1, 1.5, 0.2],
       [5. , 3.6, 1.4, 0.2],
       [5.4, 3.9, 1.7, 0.4],
       [4.6, 3.4, 1.4, 0.3],
       [5. , 3.4, 1.5, 0.2],
       [4.4, 2.9, 1.4, 0.2],
       [4.9, 3.1, 1.5, 0.1],
       [5.4, 3.7, 1.5, 0.2],
       [4.8, 3.4, 1.6, 0.2],
       [4.8, 3. , 1.4, 0.1],
       [4.3, 3. , 1.1, 0.1],
       [5.8, 4. , 1.2, 0.2],
       [5.7, 4.4, 1.5, 0.4],
       [5.4, 3.9, 1.3, 0.4],
       [5.1, 3.5, 1.4, 0.3],
       [5.7, 3.8, 1.7, 0.3],
       [5.1, 3.8, 1.5, 0.3],
       [5.4, 3.4, 1.7, 0.2],
       [5.1, 3.7, 1.5, 0.4],
       [4.6, 3.6, 1. , 0.2],
       [5.1, 3.3, 1.7, 0.5],
       [4.8, 3.4, 1.9, 0.2],
       [5. , 3. , 1.6, 0.2],
       [5. , 3.4, 1.6, 0.4],
       [5.2, 3.5, 1.5, 0.2],
       [5.2, 3.4, 1.4, 0.2],
       [4.7, 3.2, 1.6, 0.2],
       [4.8, 3.1, 1.6, 0.2],
       [5.4, 3.4, 1.5, 0.4],
       [5.2, 4.1, 1.5, 0.1],
       [5.5, 4.2, 1.4, 0.2],
       [4.9, 3

- 정답 <br>
3.6038

### 5. 붓꽃들의 각 변수별 기술통계량 구하기 : 평균, 중앙값, 표준편차

- 소수점 둘째자리까지
- 2차원 리스트의 형태로 반환
    - [[변수 1의 평균, 변수 1의 중앙값, 변수 1의 표준편차], ..., [변수 4의 평균, 변수 4의 중앙값, 변수 4의 표준편차]]

In [8]:
# def stats(X):
#     # Transpose
#     X_t = X.T

    # 밑에 코드 작성  4 X 3
    # 꽃받침 길이 , 폭 # 꽃잎 길이, 폭

# X_avg = np.mean(X_t, axis=1)
# X_median = np.median(X_t, axis=1)
# X_var = np.var(X_t, axis=1)
# v = np.vstack(X_avg, X_median, X_var)
# v = v.T

# v

def stats(X):

    # 밑에 코드 작성
    stats = np.column_stack((
        np.mean(X, axis=0), np.median(X, axis=0), np.std(X, axis=0)
    ))

    stats = np.round(stats, 2)
    return stats

result = stats(X)
result

array([[5.84, 5.8 , 0.83],
       [3.06, 3.  , 0.43],
       [3.76, 4.35, 1.76],
       [1.2 , 1.3 , 0.76]])

- 정답 <br>
[[5.84, 5.8, 0.83], [3.06, 3.0, 0.43], [3.76, 4.35, 1.76], [1.2, 1.3, 0.76]]

- numpy 공홈 : https://numpy.org/
- numpy ex100 : https://github.com/rougier/numpy-100/blob/master/100_Numpy_exercises.ipynb

In [9]:
X

array([[5.1, 3.5, 1.4, 0.2],
       [4.9, 3. , 1.4, 0.2],
       [4.7, 3.2, 1.3, 0.2],
       [4.6, 3.1, 1.5, 0.2],
       [5. , 3.6, 1.4, 0.2],
       [5.4, 3.9, 1.7, 0.4],
       [4.6, 3.4, 1.4, 0.3],
       [5. , 3.4, 1.5, 0.2],
       [4.4, 2.9, 1.4, 0.2],
       [4.9, 3.1, 1.5, 0.1],
       [5.4, 3.7, 1.5, 0.2],
       [4.8, 3.4, 1.6, 0.2],
       [4.8, 3. , 1.4, 0.1],
       [4.3, 3. , 1.1, 0.1],
       [5.8, 4. , 1.2, 0.2],
       [5.7, 4.4, 1.5, 0.4],
       [5.4, 3.9, 1.3, 0.4],
       [5.1, 3.5, 1.4, 0.3],
       [5.7, 3.8, 1.7, 0.3],
       [5.1, 3.8, 1.5, 0.3],
       [5.4, 3.4, 1.7, 0.2],
       [5.1, 3.7, 1.5, 0.4],
       [4.6, 3.6, 1. , 0.2],
       [5.1, 3.3, 1.7, 0.5],
       [4.8, 3.4, 1.9, 0.2],
       [5. , 3. , 1.6, 0.2],
       [5. , 3.4, 1.6, 0.4],
       [5.2, 3.5, 1.5, 0.2],
       [5.2, 3.4, 1.4, 0.2],
       [4.7, 3.2, 1.6, 0.2],
       [4.8, 3.1, 1.6, 0.2],
       [5.4, 3.4, 1.5, 0.4],
       [5.2, 4.1, 1.5, 0.1],
       [5.5, 4.2, 1.4, 0.2],
       [4.9, 3

가격이 10,000원인 주식이 있다. 이 주식의 일간 수익률(%)은 기댓값이 0%이고 표준편차가 1%인 표준 정규 분포를 따른다고 하자. 250일 동안의 주가를 무작위로 생성하라

In [36]:
R = np.random.randn(250)
stock = []
for i in range(250):
    price = 10000
    price = price * (R[i] / 1000 + 1)
    stock.append(price)

In [37]:
stock = np.array(stock)
print(stock)

[ 9986.37555637  9994.57935334  9998.94400692 10004.3773869
  9999.55824745 10003.6468776  10004.54848028  9998.70509296
 10002.35635421  9994.55875016  9981.89483804 10008.30485167
 10003.45196948  9998.9016658  10001.59684579 10007.53474697
 10001.96807867 10004.77214044  9985.31157878 10016.30318209
 10005.44207602 10004.45330326 10016.88475576  9990.01151387
  9995.02394025 10001.02680002 10007.12152484 10013.79074878
 10010.35029108 10012.97515124 10008.83922141  9999.16007667
 10000.89045895  9980.08828169 10009.34279759  9994.13658304
 10016.02159173 10002.5916445   9991.42164277 10007.49426147
  9997.82385429 10005.13490372 10001.76172982  9994.31244112
  9993.88121341  9988.81944224  9994.50005466  9989.56866698
 10008.28766294 10000.8228545   9989.74482138 10009.38536343
  9994.5666641  10003.56405968 10004.97431603  9977.87791219
  9990.70130676  9987.48534269 10001.89209691  9996.57965814
  9994.73282949  9997.15100953 10002.76800389 10004.6719979
 10005.0336247  10004.8401

In [57]:
R = np.random.randn(250)
stock = []
for i in range(250):
    price = 10000
    price = price * (R[i]/100 + 1)
    stock.append(price)
    
stock = np.array(stock)
print(stock)

[10067.21985974  9999.51331334  9905.9068154   9960.16262894
 10002.58711535 10036.38295983 10051.70470454  9972.88540207
  9988.8307729   9942.63196557  9933.7613961   9970.63727158
  9952.98878216  9898.57894462 10030.15316198 10117.09050883
  9953.81578234 10120.68857763  9884.25604668  9894.90662091
 10171.97918729  9977.58233002  9896.96920443  9971.38993714
  9975.25472434  9934.88774517  9841.38275289 10031.59307217
  9988.88046891 10108.17435824  9972.50072247  9880.70561299
  9838.30387351  9890.10595221 10049.81402667  9774.69170388
 10146.75453065  9974.68711841 10078.84639746 10124.37082784
 10021.94901061  9892.16432872  9747.83332599  9892.29259299
  9929.90171519 10094.81329225  9913.14437588  9991.04402243
  9946.62166466  9988.08968251 10186.02653453  9923.71630946
 10146.51178449 10069.54337827  9901.64212587  9968.66337065
  9973.37399277 10093.44501604  9879.89609725  9983.28672445
  9922.08629829 10100.0496362  10081.21346002  9989.42154525
  9883.31561972 10027.60

In [58]:
R = np.random.randn(250)
stock = []
for i in range(250):
    price = 10000
    price = price * (R[i]/100 + 1)
    stock.append(int(price))
    
stock = np.array(stock)
print(stock)

[10219  9997  9821  9954 10141 10091 10106 10123 10097 10064  9922  9961
 10036 10042 10044  9867 10034 10017 10062  9856 10245 10119 10192  9928
 10047  9879  9886 10088  9999  9966  9914 10050 10042  9883  9898 10008
  9902  9967  9859 10195 10060  9887 10231 10030  9987 10010  9945  9991
  9970  9966  9951 10050  9900  9911 10048  9818 10038  9869  9935 10098
 10031 10045 10173 10067  9890  9924  9944 10022 10009 10012 10033 10126
  9933  9985  9887  9997  9809  9816 10067  9897 10142 10147 10128  9968
  9996 10083 10051  9932 10094  9912 10159  9969 10136 10031  9973 10108
 10188 10103  9947 10054  9977  9932  9998 10064  9970  9984 10024  9966
  9923 10075  9923 10099 10227 10051  9933 10111 10039 10033 10257  9963
  9979 10071 10024  9983  9914 10063 10264  9938 10067  9899 10157  9964
  9968 10277  9926  9989 10171  9974  9809 10008  9859 10105 10075  9911
 10077  9836 10038 10152 10167  9977 10122  9825  9949 10091  9995  9885
 10035 10249  9900 10045 10117  9873 10054  9863  9

In [59]:
R = np.random.randn(250)
stock = []
price = 10000
for i in range(250):
    price = price * (R[i]/100 + 1)
    stock.append(int(price))
    
stock = np.array(stock)
print(stock)


[ 9964 10042 10053 10142 10037 10010 10101 10109 10368 10328 10350 10345
 10309 10181 10235 10556 10593 10658 10834 10844 10803 10747 10764 10830
 11025 11051 11107 11085 11154 11164 11195 11272 11246 11063 11116 11188
 11300 11346 11334 11388 11437 11416 11461 11353 11256 11245 11218 11403
 11471 11377 11321 11451 11401 11361 11420 11300 11269 11156 10943 11011
 10914 11119 11211 11251 11179 11239 11258 11409 11579 11577 11612 11564
 11511 11464 11382 11291 11340 11364 11258 11220 11260 11319 11160 11200
 11432 11141 11237 11174 11118 11115 11114 11255 11184 11260 11245 10996
 11003 10979 11076 11126 11187 11315 11174 11077 10928 11061 11085 11183
 11375 11524 11443 11449 11387 11251 11237 11151 11073 11121 11046 11164
 11231 11245 11122 11217 11105 10969 11096 10953 10774 10821 10694 10538
 10393 10374 10434 10324 10406 10288 10338 10465 10573 10560 10423 10443
 10391 10298 10353 10547 10704 10782 10738 10510 10523 10538 10673 10738
 10628 10762 10773 10932 11003 10989 10858 10864 10

In [52]:
n = 10000
x = np.random.randn(250)
for i in x:
    n = n * (1 + (i / 100))
print("{:,}원".format(round(n)))

7,188원
